### ISSUE TITLE
Misleading reportMissingTypeStubs diagnostic message and documentation

### ISSUE THREAD

**Issue by maxime-didier** [microsoft/pyright#5802]
If you have a question about a behavior that you’re seeing in pyright, consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Is your feature request related to a problem? Please describe.**
Pyright emits a `reportMissingTypeStubs` diagnostic on an import of a module with inline type annotations (no .pyi file) that is part of package with a `py.typed` marker present. The imported module is part of a company-internal package that is not fully ["type complete"](https://microsoft.github.io/pyright/#/typed-libraries?id=type-completeness) and I know of at least another dependency similarly set up (namely, result 0.13.1) that does not trigger this diagnostic when we use it.

It is unclear to me what exactly is triggering this by reading pyright's documentation. Also, the diagnostic message "Stub _file_ not found for <company.module>" is misleading since I was able to confirm via `strace` that pyright found the package's py.typed file and opened the module's .py file.

I ran `pyright --verifytypes company.module` in the project and it did find both the module and the package's py.typed file. As I expected, it also gave the module a poor completeness score and one of the functions that lack an explicit return type annotation was being called in the importing module.

**Describe the solution you’d like**
I think this is purely a documentation and error message quality issue. The diagnostic message should not tell me it did not find the stub _file_ if pyright actually found it. The [documentation](https://microsoft.github.io/pyright/#/configuration?id=type-check-diagnostics-settings) should also clarify that pyright may trigger a `reportMissingTypeStubs` when a "type incomplete" library is imported, which is what I'm guessing the actual problem is.

If that's any use, are using pyright 1.1.319.

**Comment by erictraut** [comment]
Based on your description, this sounds like a bug. You should not see a `reportMissingTypeStubs` diagnostic message if the import is targeting a package with a "py.typed" marker. The type completeness of the package shouldn't matter.

I will need more information about the issue to diagnose the problem. Here are some follow-up questions.

Is the "missing type stub" reported against the top-level module of the package or some submodule? 

Does it occur for all modules within the package or only for specific modules?

What is the directory layout of the package?

Is the package a namespace package or a traditional package (with an `__init__.py` file)?

Does the package contain combinations of namespace and traditional packages?

Is the "py.typed" marker at the top level of the package (i.e. within the top-level directory immediately beneath the `site-packages` directory)?

Is the "py.typed" marker file empty?

Are you installing the package as an editable install, or is it installed normally within your Python environment?

**Comment by maxime-didier** [comment]
Oh.

I'm getting the error against all submodules of the package which lacks an `__init__.py` file - there are no subpackages. Importing the package itself generates no diagnostic message.

The package installation is managed by nix+poetry2nix, so it is a readonly folder included in PYTHONPATH that looks like this:

```
$ tree /nix/store/(hash)-python3.10-package-name-0.1.0/lib/python3.10/site-packages
/nix/store/(hash)-python3.10-package-name-0.1.0/lib/python3.10/site-packages
├── package_name
│   ├── module_1.py
│   ├── module_2.py
│   ├── ...
│   ├── __pycache__
│   │   ├── module_1.cpython-310.pyc
│   │   ├── module_2.cpython-310.pyc
│   │   └── ...
│   └── py.typed (empty)
└── package_name-0.1.0.dist-info
    ├── direct_url.json
    ├── INSTALLER
    ├── METADATA
    ├── RECORD
    ├── REQUESTED
    └── WHEEL

3 directories, 33 files
```

Something notable about this project though: there is a json_types.py module that all the other modules depend on. It contains these definitions that seem to cause trouble in the other modules:

```python
from typing import Union
from collections.abc import Sequence, Mapping

Json = Union[None, int, float, str, 'JsonArray', 'JsonObject']
JsonArray = list[Json]
JsonObject = dict[str, Json]
JsonView = Union[None, int, float, str, 'JsonArrayView', 'JsonObjectView']
JsonArrayView = Sequence[JsonView]
JsonObjectView = Mapping[str, JsonView]
```

Using verifytypes on the json_types module, there are no warnings nor errors and I get a 100% type completeness score. In the other modules though, each time one of these types are used, I get errors such as:

```
error: Type unknown for variable "package_name.module.Class.member"
error: Type of parameter "param_name" is unknown
error: Type argument 2 for class "dict" has unknown type
```

**Comment by maxime-didier** [comment]
Here it is: https://github.com/maxime-didier/reportMissingTypeStubs-repro

**Comment by erictraut** [comment]
Thanks for putting together this example. I was able to repro the problem and fix it. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.325, which I just published. It will also be included in an upcoming pylance release.
